### PR TITLE
Pixel Template reconstruction for Phase 1

### PR DESCRIPTION
**Pull request by pmaksim1**
Template reconstruction for Phase 1, including proper treatment of the Phase 1 local coordinate systems (which violate the Run 1 convention, thereby necessitating changes to template reco).  This includes the new gen error object as well.  Requires new template payload for MC and data.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@pmaksim1 
regarding your ``` Requires new template payload for MC and data. ``` 
was this payload submitted already? 
Nominally we should test the PR together with all relevant conditions.

**Comment by pmaksim1** [comment]
Hi guys,

not sure what kind of human-level interaction is needed, but all the
relevant pixel changes are in RecoLocalTracker/SiPixelRecHits.

               Petar

**Comment by boudoul** [comment]
hi @pmaksim1 , could you please update the title of your PR ?  to be more explicit and help to keep track of the topic of the PRs  - I can suggest something like " Pixel Template reconstruction for Phase 1 "  if you wish

**Comment by pmaksim1** [comment]
Hi Slava,

@jkarancs should comment, since he tested it with the latest templates (and
everything worked fine).  This is above my git hub skill.

                   Petar

**Comment by pmaksim1** [comment]
Hi @boudol, sure!  I changed it.

            Petar

**Comment by jkarancs** [comment]
Dear all,
the new conditions (which will come together with this PR) are ready too. I already used low statistics to validate that they work qualitatively. Tomorrow after I looked at higher number of generated events, I will ask Tamas to provide AlCaDB the new conditions: LA, Templates and GenErrors. These will need to go into a new MC GT, preferably in pre5 too, but I tested already that even the old conditions will give good results (small residuals, no biases).

**Comment by civanch** [comment]
@pmaksim1 , what about run-1/2 workflows?

**Comment by jkarancs** [comment]
Dear all, It seems I found problems during the validation of the new software+conditions, which means we will need to investigate further whether it is related to this software or the db content.

File: RecoLocalTracker/SiPixelRecHits/src/SiPixelGenError.cc [hunk 1]
```diff
@@ -521,23 +522,58 @@ int SiPixelGenError::qbin(int id, float cotalpha, float cotbeta, float locBz, fl
...
-      if(locBz < 0.f) {
-         cotb = cotbeta;
-         flip_y = false;
-      } else {
-         cotb = -cotbeta;
-         flip_y = true;
-      }
-   }
+   float cota = cotalpha;
+   float cotb = acotb;;
```
Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/src/SiPixelGenError.cc]: one semicolon is spurious

File: RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc [hunk 2]
```diff
@@ -191,9 +190,9 @@ bool SiPixelTemplate::pushfile(int filenum, std::vector< SiPixelTemplateStore >
 	    theCurrentTemp.head.ss50 = theCurrentTemp.head.s50;
 	    theCurrentTemp.head.lorybias = theCurrentTemp.head.lorywidth/2.f;
 	    theCurrentTemp.head.lorxbias = theCurrentTemp.head.lorxwidth/2.f;
-	    theCurrentTemp.head.fbin[0] = fbin0;
-	    theCurrentTemp.head.fbin[1] = fbin1;
-	    theCurrentTemp.head.fbin[2] = fbin2;
+	    theCurrentTemp.head.fbin[0] = 1.5f;
+	    theCurrentTemp.head.fbin[1] = 1.00f;
+	    theCurrentTemp.head.fbin[2] = 0.85f;
```
Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc]: it's better to reuse the constants than use the magic numbers.
Please restore

File: RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc [hunk 3]
```diff
@@ -1050,19 +1053,43 @@ bool SiPixelTemplate::interpolate(int id, float cotalpha, float cotbeta, float l
...
-		} else {
-			cotb = -cotbeta;
-			flip_y = true;
-		}	
+// for some cosmics, the ususal gymnastics are incorrect  
+    cota = cotalpha; 
+    cotb = abs_cotb_;
+    flip_x = false;
+	flip_y = false;
+	switch(thePixelTemp_[index_id_].head.Dtype) {
```
Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc]: some more consistent indentation would be nice

File: RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc [hunk 4]
```diff
@@ -2485,6 +2593,52 @@ int SiPixelTemplate::qbin(int id, float cotalpha, float cotbeta, float locBz, fl
...
+    float& dy2, float& sx1, float& dx1, float& sx2, float& dx2) // float& lorywidth, float& lorxwidth)		 
+{
+	// Interpolate for a new set of track angles 
+	
+	// Local variables 
+	float locBx; //  lorywidth, lorxwidth;
+	    // Local variables 
+    locBx = 1.f;
+
+	return SiPixelTemplate::qbin(id, cotalpha, cotbeta, locBz, locBx, qclus, pixmx, sigmay, deltay, sigmax, deltax,
```
Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc]: indentation could be more consistent

**Comment by slava77** [comment]
@pmaksim1 @jkarancs 
please clarify if changes are expected in run1/run2 workflows.
From the PR description and context, I would expect no.
Instead, the comparisons show that there are changes in run1/run2 data and MC and no changes in 2017/2018 workflows (probably because of generic CPE use there)

File: RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc (hunk 4 above)
Comment (pmaksim1) [inline RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate.cc]: Hi @slava77 

just to comment that all this is in Morris's code.  It's designed to run stand-alone, and in fact this is where most of the development, testing, calibration, etc. happens.  So I'll pass this to him.  While I agree about the indentation, this is all written on a Mac so I have no idea how it shows up in there.  Moreover, I admit that all this seems exceedingly minor given the number of things we need to debug, fix, and make happen.

            Petar

**Comment by pmaksim1** [comment]
Hi @slava77 

so I'm not sure what exactly you mean by "run1/run2 workflows".  This is manifestly a run1 workflow.

Run2 detector has a different geometry.  I don't think anybody truly understands the local FPIX coordinate systems in Run2 geometry.  PixelCPEGeneric seems to run there mostly because it's generic enough to not care about x or y, as long as it's configured correctly.  But the templates are much more sophisticated, and we really need to understand E and B field in all det units before making templates that work for all sensors.  This will hopefully happen soon, but it's not ready yet.  (We have various Phase 2 templates for BPIX, but we are not sure they are correct for FPIX.)

There will be a new pull request once we know how to do Phase 2 pixel reco correctly.

Does this answer your question?  (Or was this the question?)

               Petar

**Comment by slava77** [comment]
On 2/18/17 3:50 PM, Petar Maksimovic wrote:
> Hi @slava77 <https://github.com/slava77>
> 
> so I'm not sure what exactly you mean by "run1/run2 workflows". This is
> manifestly a run1 workflow.
> 
> Run2 detector has a different geometry. I don't think anybody truly
> understands the local FPIX coordinate systems in Run2 geometry.
> PixelCPEGeneric seems to run there mostly because it's generic enough to
> not care about x or y, as long as it's configured correctly. But the
> templates are much more sophisticated, and we really need to understand
> E and B field in all det units before making templates that work for all
> sensors. This will hopefully happen soon, but it's not ready yet. (We
> have various Phase 2 templates for BPIX, but we are not sure they are
> correct for FPIX.)
> 
> There will be a new pull request once we know how to do Phase 2 pixel
> reco correctly.
> 
> Does this answer your question? (Or was this the question?)

Hi Petar,

this does not answer my question.

Tests with 2016 and run1 simulation and data show that this pull request
introduces changes in the physics outputs (tracks etc).

The title of this PR is "for phase-1"
and nothing changes in the tests using phase-1 "2017" simulation.

So, I was asking for an explanation of these observations.

Thank you.

Slava

**Comment by pmaksim1** [comment]
Hi Slava,

got it.  This is the config file(s) issue and we'll investigate.

                   Petar

File: Geometry/TrackerCommonData/data/PhaseI/trackerParameters.xml [hunk 5]
```diff
@@ -2,7 +2,7 @@
 <DDDefinition xmlns="http://www.cern.ch/cms/DDL" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.cern.ch/cms/DDL ../../../DetectorDescription/Schema/DDLSchema.xsd">
  <ConstantsSection  label="trackerParameters.xml" eval="true">
   <Vector name="vPars" type="numeric" nEntries="6">
-    80, 52, 0, 0, 2, 8
+    80, 52, 1, 2, 2, 8
```
Comment (jkarancs) [inline Geometry/TrackerCommonData/data/PhaseI/trackerParameters.xml]: Dear Ianna,
This PR was done on top of the PR that was merged already:
https://github.com/cms-sw/cmssw/pull/17384

Comment (ianna) [inline Geometry/TrackerCommonData/data/PhaseI/trackerParameters.xml]: @jkarancs - please, rebase it in that case

**Comment by ianna** [comment]
+1

**Comment by jkarancs** [comment]
Dear all,
we managed to find the problem to be related to the DB content. We would still need a few days of testing.
In order to benefit from this new software,  the full list of items to do would be the following:
- rebase and tidy up the code (including Slava's suggestions)
- create and test and validate the new DB objects and software including Phase 0 checks
- make the new DB content available in a new GT (would require another PR)
- Tracking switches on the usage of templates (another PR)

I believe pre5 (and the 90X cycle) closes too soon for the above goal, so if you agree I suggest to close this PR and continue this implementation in the upcoming 91X release cycle.

**Comment by slava77** [comment]
On 2/20/17 10:54 AM, János Karancsi wrote:
> Dear all,
> we managed to find the problem to be related to the DB content. We would
> still need a few days of testing.
> In order to benefit from this new software, the full list of items to do
> would be the following:
> 
>   * rebase and tidy up the code (including Slava's suggestions)
>   * create and test and validate the new DB objects and software
>     including Phase 0 checks
>   * make the new DB content available in a new GT (would require another PR)
>   * Tracking switches on the usage of templates (another PR)
> 
> I believe pre5 (and the 90X cycle) closes too soon for the above goal,
> so if you agree I suggest to close this PR and continue this
> implementation in the upcoming 91X release cycle.

OK for me.
We anticipate some moderate to small tracking changes in 91X cycle
beside this change.

**Comment by slava77** [comment]
-1

superseded by #17722 
moving to 91X, as discussed earlier in this thread

**Comment by pmaksim1** [comment]
Hi @slava77,

I'm going through the indentation, but basically 50% lines in the file will
move, with not a single change to code.

Is that what we want?  Just making sure before I commit something where git
diff gives 2k lines of non-changes "changes".

Thanks!

               Petar

**Comment by slava77** [comment]
On 3/4/17 8:22 AM, Petar Maksimovic wrote:
> Hi @slava77,
> 
> I'm going through the indentation, but basically 50% lines in the file will
> move, with not a single change to code.
> 
> Is that what we want? Just making sure before I commit something where git
> diff gives 2k lines of non-changes "changes".
> 
> Thanks!
> 

Hi Petar,

IIRC, my comments were local to the changed lines of code.
Some neighboring new lines in the same scope had different indentations.

I'm fine if you don't want to reindent the full file.

-- Slava

**Pull request closed without merging**